### ISSUE TITLE
Context is shared between different users

### ISSUE THREAD

**Issue by matigda** [zeit/next.js#8588]
# Bug report

## Describe the bug

When there is request to NextJS in the middle of other request, Context is being shared between those requests.

## To Reproduce

Here is repo with the code required to reproduce the bug: https://github.com/matigda/nextjs_issue

1. Clone the repo
2. Run npm install
3. Run npm run dev
4. Open http://localhost:3000 in two separate browsers ( or just open one window in private mode ) but open the second window while the first is still loading. Open your console as well before that
4b. You can also set document.cookie = 'token=whatever you wishes' ( just to see the result in console ). If you set up the cookie - refresh the page in which you didn't set up this cookie, then immediately refresh the page in second window.
5. If you didn't set up the cookie you will just see that two IDs are the same. If you set up the cookie you will also notice that token in AuthStore is the same on both pages, even though there is no cookie on one of the pages.

So basically - if request A run first and then in the middle appears request B, the A request receives data from B request. So if you run this test the other way ( meaning that you first run the window with the set up cookie and then without the cookie ) you will see that there is no token set up after page is loaded, even though the cookie is there. 

## Expected behavior

Well...I think it's obious that it's not desired in this case to share those objects between users. But I'm not sure whether it's not just a bug in my "withMobxStores" container, so please take a look there first.

## System information

- OS: Linux
- Version of Next.js: tested on 8.1.0 and 9.0.5

## Additional context

Please let me know if this description is enough. If not - I will record it and share.

**Comment by matigda** [comment]
Yeah, ok, it's due to closure in stores/index.js. But it's strange behaviour. I would understand if this instance is shared amongst all the users, but in this case it seems as if it's shared only if the previous request was not ended. Can someone exaplain a little bit more on how processes and variables work in NextJS ( server side ) ? Or is it express thing?

**Comment by rande** [comment]
There is no error in nextjs, a module is a singleton for the current thread. so any variable created in that scope is shared. So just don't create a context in a global scope.

**Comment by Timer** [comment]
I'm closing this because general usage questions belong on our Spectrum or Stack Overflow.

This is not a bug with Next.js, as this is standard Node behavior. Please adjust your application to scope any global data to the request-level.

**Comment by JoaoFGuiomar** [comment]
@matigda Did you solve this? Think i'm having a similar issue

**Comment by matigda** [comment]
Yep, I just don't create global variable, as said here.
It is a little confusing as example here: https://github.com/zeit/next.js/blob/canary/examples/with-mobx-react-lite/store.js
is, from what I understand, vurnelable to the same thing.

**Comment by JoaoFGuiomar** [comment]
Thank you.

 I have the same “symptoms” on our project. Its very bizarre. Specially the “time” aspect of it.

The project is getting bigger and bigger, its going to be interesting to find this global variable 😫

**Comment by Dhonatan** [comment]
Hi guys, how is it going?

I have the exactly same problem. NEXT + ReactJS SSR.

I have my CONTEXT with the token of my users, and when I need to make a request in the API, I use the TOKEN from the CONTEXT... so the CONTEXT is global (because I need to use in a lot of pages).

What's the better solution for the problem?

Remove the CONTEXT and save this TOKEN on STORE/COOKIE and when I need to make a request I catch it from there?

I use this CONTEXT in my footer/header, to show a different type of menu from the user (if its logged in or not). Whats should I do on it? because if I catch from the store/cookie every time that the user move from the one page to another, will not be 'fast'.

Thanks

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
